### ISSUE TITLE
Performance issue with the method Members.MemberHasAccess(Path)

### ISSUE THREAD

**Issue by facingdigital** [umbraco/Umbraco-CMS#7292]
Hi,

We have noticed a performance issue on the method Members.MemberHasAccess(Path) if it has been used as a filter in a query with Umbraco 8.3 containing more than 5000 documents/pages.

The loading time is multiplied per 10 if this method is used in the filter linked to the number of access to check.

Could you please check and improve the performance of this method ?

Thank you.

---
_This item has been added to our backlog [AB#7652](https://dev.azure.com/umbraco/243e7927-03b2-44e2-908f-d4ac7ea5daaa/_workitems/edit/7652)_

**Comment by nul800sebastiaan** [comment]
Hey there @facingdigital. Thanks for the report. We need a bit more information from you:

- From what I can tell from a quick look at the code, each `MemberHasAccess` call is cached (the cache key is the path) after the first call - are subsequent calls of this method also slow for you?
- I don't understand what you mean by "used as a filter in a query with Umbraco 8.3 containing more than 5000 documents/pages"?
- I don't understand what you mean by "multiplied per 10 if this method is used in the filter linked to the number of access to check"?

**Comment by facingdigital** [comment]
Hi @nul800sebastiaan,

Sorry if my description is not clear :)

We use the method "Members.MemberHasAccess(x.Path)" inside a linq query to filter the children pages according to the Public Access.

For example, we create list of children page thanks a query as :

Model.Children.Where(x=>Members.MemberHasAccess(x.Path))
But Members.MemberHasAccess(x.Path) causes an big impact on the performance. 

To test it, it's easy, you must create 2 content types, the parent and one child pages (Events and event for example) Then you create, more than 100 event pages.

If you execute a query without the condition Members.MemberHasAccess(x.Path), the page will be loaded quickly... Then, you add the condition and load the page several time... 

It's important to load several time the page, because we have noticed that more the page has been loaded and more the page takes time to load on our Umbraco Cloud Pro (live environment)

[Example on our site](https://www.scienceeurope.org/our-resources/)

Currently, the page is quick because we have found a work around thanks a condition on our property of the content type containing the member groups authorized to see the "resource".
(we have developed an Event Composer in charge to set the public access according to the selected member groups of our data property, it's easier for the contributor and we can direct filter the items according to the Public Access values.)
 
But it's strange, that Members.MemberHasAccess causes a big impact on the loading time of the page.

**We have noticed that without using Members.MemberHasAccess, the memory usage had decreased.** 
Before, the consumed memory could be more than 800Mb and now, the average is around 400MB...

Thank you for your support

**Comment by Shazwazza** [comment]
I can confirm there is an issue here. In order to replicate you need to have a bunch of checks for Members.MemberHasAccess(x.Path) with different paths and be logged in as a member and have most of those paths protected by public access in the back office.

In that case there's a query that is not cached: `MemberGroupRepository.GetMemberGroupsForMember` which will execute each time. Which of course could mean quite a lot of SQL calls.

A current work around would be to only call `Members.MemberHasAccess(x.Path)` on paths that don't overlap. For example, you could get all the paths you want to check and order them ascending, then check from the upper most path for the member and you'll know you don't have to check the lower paths if you have access to the upper path.

I will look at a fix for this internally .

**Comment by Shazwazza** [comment]
There's a band-aid fix for this which would be to cache the results of `MemberGroupRepository.GetMemberGroupsForMember` but that will still incur a SQL (very fast) query each time because we need to acquire a readlock (distributed lock). Adding this cache would massively improve this but I'd prefer that we still aren't making readlock SQL calls either.

I think the correct fix would be to have a method that checks for multiple path access at once like 

```cs
IDictionary<string, bool> Members.MemberHasAccess(IEnumerable<string> paths);
```

Which would return a dictionary of paths and if the member has access

**Comment by Shazwazza** [comment]
As for the memory consumption, the only thing I can conclude there is that there is already a band-aid fix in place for this method that caches the result of each call to this within the request. Now if you have TONS of calls to this method, then each request size is going to be large and if lots of requests are made then this memory is going to get put on the Gen2 space for garbage collection which will remain their longer than normal and will take the GC some time to cleanup... which means memory grows faster, will eventually be cleaned but not if there are still requests coming in for this code.

**Comment by Shazwazza** [comment]
Have created a PR and solution/description here https://github.com/umbraco/Umbraco-CMS/pull/7596

**Comment by clausjensen** [comment]
The newly added option for checking in bulk should make it possible to avoid much of the performance problem involved in checking access on multiple paths now. This is planned to be released in Umbraco CMS 8.8.0 - I'll close this issue for now.
